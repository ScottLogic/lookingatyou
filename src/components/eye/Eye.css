svg.layer {
    position: absolute;
    width: 50%;
    height: 100%;
}

.LEFT,
.RIGHT {
    display: inline-block;
    height: 100%;
    width: 50%;
}

svg.overlay {
    transform-origin: center;
    position: absolute;
}

div.container {
    height: 100%;
    width: 100%;
}

.LEFT,
.RIGHT {
    display: inline-block;
    height: 100%;
    width: 50%;
}
