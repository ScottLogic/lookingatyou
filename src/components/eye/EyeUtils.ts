import {
    buffer,
    dilationMultipler,
    dilationOffset,
    irisSkewFactor,
    maxBrightness,
    middleX,
    moveSize,
    xIncrement,
} from '../../AppConstants';

export function analyseLight(
    image: ImageData,
    tooBright: boolean,
): { tooBright: boolean; scaledPupilSize: number } {
    if (image.width > 0) {
        const data = image.data;
        let colorSum = 0;
        for (let i = 0; i < data.length; i += 4) {
            const avg = Math.floor((data[i] + data[i + 1] + data[i + 2]) / 3);

            colorSum += avg;
        }
        let brightness = Math.floor(colorSum / (image.width * image.height));

        if (brightness > maxBrightness) {
            tooBright = true;
            brightness = maxBrightness;
        } else if (tooBright) {
            tooBright = false;
        }
        const scaledPupilSize =
            ((maxBrightness - brightness) / maxBrightness) * dilationMultipler +
            dilationOffset;

        return { tooBright, scaledPupilSize };
    }
    return { tooBright: false, scaledPupilSize: 0 };
}

export function naturalMovement(
    currentX: number,
    isMovingLeft: boolean,
): { newX: number; isMovingLeft: boolean } {
    if (currentX === middleX) {
        if (Math.random() < 0.1) {
            return moveEye(currentX, isMovingLeft);
        }
        return { newX: 0, isMovingLeft };
    } else {
        return moveEye(currentX, isMovingLeft);
    }
}

function moveEye(
    currentX: number,
    isMovingLeft: boolean,
): { newX: number; isMovingLeft: boolean } {
    if (isMovingLeft) {
        return moveLeft(currentX, isMovingLeft);
    } else {
        return moveRight(currentX, isMovingLeft);
    }
}

function moveLeft(
    currentX: number,
    isMovingLeft: boolean,
): { newX: number; isMovingLeft: boolean } {
    if (currentX > middleX - xIncrement + buffer) {
        return { newX: currentX - moveSize, isMovingLeft };
    } else if (Math.random() < 0.5) {
        return { newX: currentX + moveSize, isMovingLeft: !isMovingLeft };
    }
    return { newX: currentX, isMovingLeft };
}

function moveRight(
    currentX: number,
    isMovingLeft: boolean,
): { newX: number; isMovingLeft: boolean } {
    if (currentX < middleX + xIncrement - buffer) {
        return { newX: currentX + moveSize, isMovingLeft };
    } else if (Math.random() < 0.5) {
        return { newX: currentX - moveSize, isMovingLeft: !isMovingLeft };
    }
    return { newX: currentX, isMovingLeft };
}

export function getMaxDisplacement(scleraRadius: number, irisRadius: number) {
    return (scleraRadius - irisRadius * irisSkewFactor) / irisSkewFactor;
}

export function getIrisAdjustment(
    x: number,
    y: number,
    height: number,
    width: number,
    scleraRadius: number,
    irisRadius: number,
    previousAngle: number = 0,
) {
    const displacement = Math.hypot(x - width / 2, y - height / 2);
    const maxDisplacement = getMaxDisplacement(scleraRadius, irisRadius);

    const scale =
        irisSkewFactor +
        ((1 - irisSkewFactor) * (maxDisplacement - displacement)) /
            maxDisplacement;

    let angle = (Math.atan2(y - height / 2, x - width / 2) * 180) / Math.PI;

    if (angle - previousAngle < -90) {
        angle = angle + 180;
    } else if (angle - previousAngle > 90) {
        angle = angle - 180;
    }

    return {
        scale,
        angle,
    };
}
