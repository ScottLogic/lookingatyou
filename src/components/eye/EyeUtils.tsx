import {
    buffer,
    dilationMultipler,
    dilationOffset,
    maxBrightness,
    middleX,
    moveSize,
    xIncrement,
} from '../../AppConstants';

export function checkLight(
    tooBright: boolean,
    video:
        | ImageData
        | HTMLImageElement
        | HTMLCanvasElement
        | HTMLVideoElement
        | null,
    callback: (
        canvas: HTMLCanvasElement,
        tooBright: boolean,
    ) => [boolean, number],
): [boolean, number] {
    if (video && video instanceof HTMLVideoElement) {
        const canvas = document.createElement('canvas');
        canvas.height = video.height;
        canvas.width = video.width;
        const canvasCtx = canvas.getContext('2d');

        if (canvasCtx) {
            canvasCtx.drawImage(video, 0, 0, canvas.width, canvas.height);
        }
        return callback(canvas, tooBright);
    }
    return [false, 0];
}

export function analyseLight(
    canvas: HTMLCanvasElement,
    tooBright: boolean,
): [boolean, number] {
    const ctx = canvas.getContext('2d');

    if (ctx && canvas.width > 0) {
        const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);

        const data = imageData.data;

        let colorSum = 0;

        for (let i = 0; i < data.length; i += 4) {
            const avg = Math.floor((data[i] + data[i + 1] + data[i + 2]) / 3);

            colorSum += avg;
        }

        let brightness = Math.floor(colorSum / (canvas.width * canvas.height));

        if (brightness > maxBrightness) {
            tooBright = true;
            brightness = maxBrightness;
        } else if (tooBright) {
            tooBright = false;
        }

        const scaledPupilSize =
            ((maxBrightness - brightness) / maxBrightness) * dilationMultipler +
            dilationOffset;

        return [tooBright, scaledPupilSize];
    }
    return [false, 0];
}

export function naturalMovement(
    currentX: number,
    direction: boolean,
): [number, boolean] {
    if (currentX === middleX) {
        if (Math.random() < 0.01) {
            return moveEye(currentX, direction);
        }
        return [0, direction];
    } else {
        return moveEye(currentX, direction);
    }
}

function moveEye(currentX: number, direction: boolean): [number, boolean] {
    if (direction) {
        return moveLeft(currentX, direction);
    } else {
        return moveRight(currentX, direction);
    }
}

function moveLeft(currentX: number, direction: boolean): [number, boolean] {
    if (currentX > middleX - xIncrement + buffer) {
        return [currentX - moveSize, direction];
    } else if (Math.random() < 0.1) {
        return [currentX, !direction];
    }
    return [currentX, direction];
}

function moveRight(currentX: number, direction: boolean): [number, boolean] {
    if (currentX < middleX + xIncrement - buffer) {
        return [currentX + moveSize, direction];
    } else if (Math.random() < 0.1) {
        return [currentX, !direction];
    }
    return [currentX, direction];
}
