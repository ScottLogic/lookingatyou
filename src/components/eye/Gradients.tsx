import React from 'react';
import tinycolor from 'tinycolor2';
import './Gradients.css';

interface IGradientsProps {
    pupilColor: string;
    irisColor: string;
}

const lighterFactor = 30;
const darkerFactor = 13;

export function Gradients(props: IGradientsProps) {
    const darkIrisColor = tinycolor(props.irisColor)
        .darken(darkerFactor)
        .toHexString();
    const lightIrisColor = tinycolor(props.irisColor)
        .brighten(lighterFactor)
        .toHexString();

    return (
        <svg className={'collapse'}>
            <defs>
                <linearGradient id="lgrad" x1="0%" y1="50%" x2="100%" y2="50%">
                    <stop offset="35%" stopColor={props.irisColor} />
                    <stop offset="70%" stopColor={lightIrisColor} />
                    <stop offset="100%" stopColor={darkIrisColor} />
                </linearGradient>

                <radialGradient id={'scleraGradient'} cx={'50%'} cy={'50%'}>
                    <stop offset={'0%'} stopColor={'white'} />
                    <stop
                        offset={'85%'}
                        stopColor={tinycolor('white')
                            .darken(10)
                            .toHexString()}
                    />
                    <stop
                        offset={'100%'}
                        stopColor={tinycolor('white')
                            .darken(30)
                            .toHexString()}
                    />
                </radialGradient>

                <radialGradient
                    id="irisGradient"
                    cx={'50%'}
                    cy={'50%'}
                    href={'#lgrad'}
                />

                <radialGradient id="reflectionGradient">
                    <stop offset="0%" stopColor={'white'} stopOpacity={1} />
                    <stop offset="20%" stopColor={'white'} stopOpacity={0.7} />
                    <stop offset="40%" stopColor={'white'} stopOpacity={0.05} />
                </radialGradient>
            </defs>
        </svg>
    );
}
