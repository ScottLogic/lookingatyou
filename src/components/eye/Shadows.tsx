import React from 'react';
import { eyelidPosition } from '../../AppConstants';

interface IShadowsProps {
    openCoefficient: number;
}

export function Shadows(props: IShadowsProps) {
    return (
        <svg>
            <defs>
                <filter id="shadowTop">
                    <feDropShadow
                        dx="0"
                        dy="20"
                        stdDeviation="20"
                        floodColor="#333333"
                    />
                </filter>
                <filter id="shadowBottom">
                    <feDropShadow
                        dx="0"
                        dy={
                            props.openCoefficient === eyelidPosition.CLOSED
                                ? 10
                                : -5
                        }
                        stdDeviation="15"
                        floodColor="#333333"
                    />
                </filter>
            </defs>
        </svg>
    );
}
