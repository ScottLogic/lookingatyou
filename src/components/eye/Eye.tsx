import React from 'react';
import { eyes, transitionTime } from '../../AppConstants';
import './Eye.css';

interface IEyeProps {
    class: string;
    width: number;
    height: number;
    irisColor: string;
    scleraRadius: number;
    irisRadius: number;
    pupilRadius: number;
    openCoefficient: number;
    dilatedCoefficient: number;
    innerX: number;
    innerY: number;
    fps: number;
}

export default class Eye extends React.Component<IEyeProps> {
    private circleTransitionStyle: { transition: string };
    private eyelidTransitionStyle: { transition: string };
    constructor(props: IEyeProps) {
        super(props);
        this.circleTransitionStyle = {
            transition: `r ${transitionTime.dilate}ms, cx ${1000 /
                props.fps}ms, cy ${1000 / props.fps}ms`, // cx and cy transitions based on FPS
        };
        this.eyelidTransitionStyle = {
            transition: `d ${transitionTime.blink}ms`,
        };
    }

    renderCircle(
        radius: number,
        name: string,
        colour: string,
        centerX: number = this.props.width / 2,
        centerY: number = this.props.height / 2,
    ) {
        return (
            <circle
                style={this.circleTransitionStyle}
                r={radius}
                className={name}
                fill={colour}
                cx={centerX}
                cy={centerY}
            />
        );
    }

    render() {
        const eyeMiddleX = this.props.width / 2;
        const eyeLeft = eyeMiddleX - this.props.scleraRadius;
        const eyeRight = eyeMiddleX + this.props.scleraRadius;
        const eyeMiddleY = this.props.height / 2;

        const topEyelidY =
            eyeMiddleY - this.props.scleraRadius * this.props.openCoefficient;
        const bottomEyelidY =
            eyeMiddleY + this.props.scleraRadius * this.props.openCoefficient;

        const bezierCurveConstant = 0.55228474983; // (4/3)tan(pi/8)
        const bezierControlOffset =
            this.props.scleraRadius * bezierCurveConstant;
        const scaledYBezierControlOffset =
            bezierControlOffset * this.props.openCoefficient;
        const scaledXBezierControlOffset =
            bezierControlOffset - scaledYBezierControlOffset;
        const innerTopCoefficient = 1.25;
        const innerBottomCoefficient = 1.1;
        const outerTopCoefficient = 0.85;
        const outerBottomCoefficient = 0.9;
        let leftTopCoefficient;
        let rightTopCoefficient;
        let leftBottomCoefficient;
        let rightBottomCoefficient;
        if (this.props.class === eyes.RIGHT) {
            leftTopCoefficient = innerTopCoefficient;
            rightTopCoefficient = outerTopCoefficient;
            leftBottomCoefficient = innerBottomCoefficient;
            rightBottomCoefficient = outerBottomCoefficient;
        } else {
            leftTopCoefficient = outerTopCoefficient;
            rightTopCoefficient = innerTopCoefficient;
            leftBottomCoefficient = outerBottomCoefficient;
            rightBottomCoefficient = innerBottomCoefficient;
        }
        return (
            <svg
                className={this.props.class}
                width={this.props.width}
                height={this.props.height}
            >
                {this.renderCircle(this.props.scleraRadius, 'sclera', 'white')}
                <g className="inner">
                    {this.renderCircle(
                        this.props.irisRadius,
                        'iris',
                        this.props.irisColor,
                        this.props.innerX,
                        this.props.innerY,
                    )}
                    {this.renderCircle(
                        this.props.pupilRadius * this.props.dilatedCoefficient,
                        'pupil',
                        'black',
                        this.props.innerX,
                        this.props.innerY,
                    )}
                </g>
                <svg className="Eyelids">
                    <path
                        style={this.eyelidTransitionStyle}
                        d={
                            // upper eyelid
                            `M ${eyeLeft} ${eyeMiddleY},
                         A ${this.props.scleraRadius} ${
                                this.props.scleraRadius
                            } 0 0 1 ${eyeRight} ${eyeMiddleY}
                         C ${eyeRight -
                             rightTopCoefficient *
                                 scaledXBezierControlOffset} ${eyeMiddleY -
                                scaledYBezierControlOffset}, 
                         ${eyeMiddleX +
                             bezierControlOffset} ${topEyelidY}, ${eyeMiddleX} ${topEyelidY}
                         C ${eyeMiddleX -
                             bezierControlOffset} ${topEyelidY}, ${eyeLeft +
                                leftTopCoefficient *
                                    scaledXBezierControlOffset} ${eyeMiddleY -
                                scaledYBezierControlOffset}, ${eyeLeft} ${eyeMiddleY}`
                        }
                    />
                    <path
                        style={this.eyelidTransitionStyle}
                        d={
                            // lower eyelid
                            `M ${eyeLeft} ${eyeMiddleY},
                         A ${this.props.scleraRadius} ${
                                this.props.scleraRadius
                            } 0 0 0 ${eyeRight} ${eyeMiddleY}
                         C ${eyeRight -
                             rightBottomCoefficient *
                                 scaledXBezierControlOffset} ${eyeMiddleY +
                                scaledYBezierControlOffset}, 
                         ${eyeMiddleX +
                             bezierControlOffset} ${bottomEyelidY}, ${eyeMiddleX} ${bottomEyelidY}
                         C ${eyeMiddleX -
                             bezierControlOffset} ${bottomEyelidY}, ${eyeLeft +
                                leftBottomCoefficient *
                                    scaledXBezierControlOffset} ${eyeMiddleY +
                                scaledYBezierControlOffset}, ${eyeLeft} ${eyeMiddleY}`
                        }
                    />
                </svg>
                <svg className="BlackFill">
                    <path
                        d={
                            // upper eyelid
                            `M ${0} ${eyeMiddleY},
                         L ${eyeLeft} ${eyeMiddleY},
                         A ${this.props.scleraRadius} ${
                                this.props.scleraRadius
                            } 0 0 1 ${eyeRight} ${eyeMiddleY}
                         L ${this.props.width} ${eyeMiddleY},
                         L ${this.props.width} 0
                         L 0 0`
                        }
                    />
                    <path
                        d={
                            // upper eyelid
                            `M ${0} ${eyeMiddleY},
                         L ${eyeLeft} ${eyeMiddleY},
                         A ${this.props.scleraRadius} ${
                                this.props.scleraRadius
                            } 0 0 0 ${eyeRight} ${eyeMiddleY}
                         L ${this.props.width} ${eyeMiddleY},
                         L ${this.props.width} ${this.props.height}
                         L 0 ${this.props.height}`
                        }
                    />
                </svg>
            </svg>
        );
    }
}
