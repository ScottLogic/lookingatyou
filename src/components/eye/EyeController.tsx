import React, { useEffect, useRef, useState } from 'react';
import isEqual from 'react-fast-compare';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import {
    blinkConsts,
    eyeRadiiCoefficients,
    EyeSide,
    minIrisScale,
    numInnerEyeSectors,
    transitionTimes,
} from '../../AppConstants';
import { IDetection } from '../../models/objectDetection';
import { IConfigState } from '../../store/actions/config/types';
import { setAnimation } from '../../store/actions/detections/actions';
import { ISetAnimationAction } from '../../store/actions/detections/types';
import { IRootStore } from '../../store/reducers/rootReducer';
import { getConfig } from '../../store/selectors/configSelectors';
import {
    getSelections,
    getTargets,
} from '../../store/selectors/detectionSelectors';
import { getVideo } from '../../store/selectors/videoSelectors';
import { Animation, blink } from '../../utils/pose/animations';
import { ICoords } from '../../utils/types';
import Eye from './Eye';
import { Gradients } from './Gradients';
import { Shadows } from './Shadows';
import { getReflection } from './utils/ReflectionUtils';
import { generateInnerPath, irisMatrixTransform } from './utils/VisualUtils';

interface IEyeControllerProps {
    width: number;
    height: number;
    environment: Window;
    dilation: number;
    detected: boolean;
    openCoefficient: number;
}

interface IEyeControllerMapStateToProps {
    config: IConfigState;
    target: ICoords;
    image: HTMLVideoElement | undefined;
    selection: IDetection | undefined;
    animation: Animation;
}

interface IEyeControllerMapDispatchToState {
    updateAnimation: (animation: Animation) => ISetAnimationAction;
}

export type EyeControllerProps = IEyeControllerProps &
    IEyeControllerMapStateToProps &
    IEyeControllerMapDispatchToState;

export const EyeController = React.memo(
    (props: EyeControllerProps) => {
        const { environment, updateAnimation, animation } = props;

        const scleraRadius = Math.floor(
            props.width * eyeRadiiCoefficients.sclera,
        );
        const irisRadius = Math.floor(props.width * eyeRadiiCoefficients.iris);
        const pupilRadius = Math.floor(
            props.width * eyeRadiiCoefficients.pupil,
        );

        const irisAdjustmentRef = useRef({
            scale: 1,
            angle: 0,
        });

        const reflectionRef = useRef<ImageData | undefined>(undefined);

        const { x, y } =
            props.animation.length > 0 &&
            props.animation[0].normalisedCoords !== undefined
                ? {
                      x: props.animation[0].normalisedCoords!.x,
                      y: props.animation[0].normalisedCoords!.y,
                  }
                : { x: props.target.x, y: props.target.y };

        const maxDisplacement =
            (scleraRadius - irisRadius * minIrisScale) / minIrisScale;
        const targetY = y * props.config.ySensitivity;
        const targetX = -x * props.config.xSensitivity; // mirrored
        const polarDistance = Math.hypot(targetY, targetX);
        const polarAngle = Math.atan2(targetY, targetX);
        const displacement = Math.min(1, polarDistance) * maxDisplacement;
        const innerX = props.width / 4 + displacement * Math.cos(polarAngle);
        const innerY = props.height / 2 + displacement * Math.sin(polarAngle);

        const calculatedEyesOpenCoefficient =
            props.animation.length > 0 &&
            props.animation[0].hasOwnProperty('openCoefficient')
                ? props.animation[0].openCoefficient!
                : props.openCoefficient;

        const dilatedCoefficient =
            props.animation.length > 0 &&
            props.animation[0].dilation !== undefined
                ? props.animation[0].dilation
                : props.dilation;

        const [innerPath, setInnerPath] = useState(
            generateInnerPath(irisRadius, numInnerEyeSectors),
        );

        const irisColor =
            props.animation.length > 0 && props.animation[0].irisColor
                ? props.animation[0].irisColor
                : props.config.irisColor;

        const animationRef = useRef(animation);
        const detectedRef = useRef(props.detected);
        useEffect(() => {
            detectedRef.current = props.detected;
        }, [props.detected]);
        useEffect(() => {
            animationRef.current = animation;
        }, [animation]);

        useEffect(() => {
            if (animationRef.current.length === 0) {
                let blinkInterval = environment.setInterval(() => {
                    const blinkFrequency = detectedRef.current
                        ? blinkConsts.focusedFrequency
                        : blinkConsts.frequency;
                    const blinkProbability =
                        blinkFrequency / (1000 / transitionTimes.blink);
                    if (Math.random() < blinkProbability) {
                        updateAnimation(blink());
                    }
                }, transitionTimes.blink);
                return () => {
                    environment.clearInterval(blinkInterval);
                    blinkInterval = 0;
                };
            }
        }, [environment, updateAnimation]);

        useEffect(() => {
            if (animation.length > 0) {
                const timer = environment.setTimeout(() => {
                    const myAnimation = [...animation];
                    myAnimation!.shift();
                    updateAnimation(myAnimation);
                }, animation[0].duration);
                return () => environment.clearTimeout(timer);
            }
        }, [animation, updateAnimation, environment]);

        useEffect(() => {
            if (
                props.config.toggleReflection &&
                props.selection &&
                props.image
            ) {
                reflectionRef.current = getReflection(
                    pupilRadius,
                    props.selection.bbox,
                    props.image,
                );
            } else {
                reflectionRef.current = undefined;
            }
        }, [
            props.selection,
            props.image,
            props.config.toggleReflection,
            pupilRadius,
        ]);

        useEffect(() => {
            setInnerPath(generateInnerPath(irisRadius, 100));
        }, [irisRadius]);

        return (
            <div className="container">
                {[EyeSide.RIGHT, EyeSide.LEFT].map((eye, index) => {
                    const bezier = getBezier(
                        scleraRadius,
                        typeof calculatedEyesOpenCoefficient === 'number'
                            ? calculatedEyesOpenCoefficient
                            : calculatedEyesOpenCoefficient[eye],
                    );

                    const eyeShape = getEyeShape(
                        props.width,
                        props.height,
                        scleraRadius,
                        typeof calculatedEyesOpenCoefficient === 'number'
                            ? calculatedEyesOpenCoefficient
                            : calculatedEyesOpenCoefficient[eye],
                    );
                    return (
                        <Eye
                            {...props}
                            class={eye}
                            key={index}
                            width={props.width / 2}
                            irisColor={irisColor}
                            scleraRadius={scleraRadius}
                            irisRadius={irisRadius}
                            pupilRadius={pupilRadius}
                            dilatedCoefficient={dilatedCoefficient}
                            innerX={innerX}
                            innerY={innerY}
                            fps={props.config.fps}
                            bezier={bezier}
                            eyeShape={eyeShape}
                            reflection={reflectionRef.current}
                            irisAdjustment={irisAdjustmentRef.current}
                            innerPath={innerPath}
                            skewTransform={irisMatrixTransform(props.target)}
                            transformDuration={
                                props.animation.length > 0
                                    ? props.animation[0].duration
                                    : undefined
                            }
                        />
                    );
                })}
                <Gradients
                    irisColor={irisColor}
                    reflectionOpacity={props.config.reflectionOpacity}
                />
                <Shadows openCoefficient={props.openCoefficient} />
            </div>
        );
    },
    (previous, next) =>
        isEqual(previous.animation, next.animation) &&
        previous.dilation === next.dilation &&
        previous.openCoefficient === next.openCoefficient &&
        previous.target.x === next.target.x &&
        previous.target.y === next.target.y,
);

export function getBezier(scleraRadius: number, openCoefficient: number) {
    const curveConstant = 0.55228474983; // (4/3)tan(pi/8)
    const controlOffset = scleraRadius * curveConstant;
    const scaledYcontrolOffset = controlOffset * openCoefficient;
    const scaledXcontrolOffset = controlOffset - scaledYcontrolOffset;
    return { controlOffset, scaledXcontrolOffset, scaledYcontrolOffset };
}

export function getEyeShape(
    width: number,
    height: number,
    scleraRadius: number,
    openCoefficient: number,
) {
    const middleX = width / 4;
    const leftX = middleX - scleraRadius;
    const rightX = middleX + scleraRadius;
    const middleY = height / 2;

    const topEyelidY = middleY - scleraRadius * openCoefficient;
    const bottomEyelidY = middleY + scleraRadius * openCoefficient;

    return {
        middleX,
        leftX,
        rightX,
        middleY,
        topEyelidY,
        bottomEyelidY,
    };
}

const mapStateToProps = (state: IRootStore): IEyeControllerMapStateToProps => ({
    config: getConfig(state),
    target: getTargets(state),
    image: getVideo(state),
    selection: getSelections(state),
    animation: state.detectionStore.animation,
});

const mapDispatchToProps = (
    dispatch: Dispatch,
): IEyeControllerMapDispatchToState => ({
    updateAnimation: (animation: Animation) =>
        dispatch(setAnimation(animation)),
});

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(EyeController);
