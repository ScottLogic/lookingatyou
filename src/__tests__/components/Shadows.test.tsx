import { shallow } from 'enzyme';
import React from 'react';
import { IShadowsProps, Shadows } from '../../components/eye/Shadows';

let props: IShadowsProps;

describe('Eye Controller', () => {
    beforeEach(() => {
        props = {
            openCoefficient: 1,
        };
    });

    it('should render correctly', () => {
        const wrapper = shallow(<Shadows {...props} />).debug();
        expect(wrapper).toMatchSnapshot();
    });
});
