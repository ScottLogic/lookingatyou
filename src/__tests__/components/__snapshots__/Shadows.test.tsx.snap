// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shadows should render correctly 1`] = `
"<svg className=\\"collapse\\">
  <defs>
    <filter id=\\"shadowTop\\">
      <feDropShadow dx=\\"0\\" dy={44.44444444444444} stdDeviation={44.44444444444444} floodColor=\\"#333333\\" />
    </filter>
    <filter id=\\"shadowBottom\\">
      <feDropShadow dx=\\"0\\" dy={-11.11111111111111} stdDeviation={33.333333333333336} floodColor=\\"#333333\\" />
    </filter>
  </defs>
</svg>"
`;
